crates/nu-command/src/formats/to/msgpack.rs

```diff
@@ -208,25 +228,53 @@ pub(crate) fn write_value(
             mp::write_map_len(out, convert(val.len(), span)?).err_span(span)?;
             for (k, v) in val.iter() {
                 mp::write_str(out, k).err_span(span)?;
-                write_value(out, v, depth + 1)?;
+                write_value(out, v, depth + 1, engine_state, call_span, serialize_types)?;
             }
         }
         Value::List { vals, .. } => {
             mp::write_array_len(out, convert(vals.len(), span)?).err_span(span)?;
             for val in vals {
-                write_value(out, val, depth + 1)?;
+                write_value(
+                    out,
+                    val,
+                    depth + 1,
+                    engine_state,
+                    call_span,
+                    serialize_types,
+                )?;
             }
         }
         Value::Nothing { .. } => {
             mp::write_nil(out)
                 .map_err(InvalidMarkerWrite)
                 .err_span(span)?;
         }
-        Value::Closure { .. } => {
-            // Closures can't be converted
-            mp::write_nil(out)
-                .map_err(InvalidMarkerWrite)
-                .err_span(span)?;
+        Value::Closure { val, .. } => {
+            if serialize_types {
+                let block = engine_state.get_block(val.block_id);
+                let closure_string: &str = if let Some(span) = block.span {
+                    let contents_bytes = engine_state.get_span_contents(span);
+                    &String::from_utf8_lossy(contents_bytes)
+                } else {
+                    return Err(WriteError::Shell(Box::new(ShellError::CantConvert {
+                        to_type: "string".into(),
+                        from_type: "closure".into(),
+                        span,
+                        help: Some(format!(
+                            "unable to retrieve block contents for closure with id {}",
+                            val.block_id.get()
+                        )),
+                    })));
+                };
```

Comment:
Could this part you repeated across a few places live on the closure type?

e.g. `Closure::coerce_to_string`

Reply:
that sounds good to me, I thought of doing something like that but wasn't sure where to put it. not sure if `Closure` is really the _best_ place to put it, but it keeps it out of the other stringifying logic so it's good enough for me. we can probably revisit this as part of #15286 and #14090

Reply:
I like on the type more than in a large `match` like on `Value`. Could also be useful if you want to debug print stuff in places like the engine.

With engine_state should be fine in `nu-protocol`